release-22.1: colexec: memory accounting at row granularity in columnarizer

yuzefovich opened the pull request:
Backport 2/2 commits from #87242.
Backport 1/1 commits from #87471.

/cc @cockroachdb/release

---

**colexec: remove redundant allocations when converting from rows**

This redundant allocation has been present since the introduction of the
vectorized engine (it was introduced in
37957a9b4e6964940ed2dd4d94b4dbfa671a11aa I think by accident - the
impact of some changes in that commit were probably not understood and
we didn't have the benchmarks to catch the regression).
```
Columnarize-24                442µs ± 1%     157µs ± 0%   -64.49%  (p=0.000 n=10+9)
ColumnarizeMaterialize-24     801µs ± 1%     513µs ± 0%   -35.89%  (p=0.000 n=10+10)

name                       old speed      new speed      delta
Columnarize-24              362MB/s ± 1%  1019MB/s ± 0%  +181.61%  (p=0.000 n=10+9)
ColumnarizeMaterialize-24   200MB/s ± 1%   312MB/s ± 0%   +55.97%  (p=0.000 n=10+10)

name                       old alloc/op   new alloc/op   delta
Columnarize-24                157kB ± 0%       0kB ± 1%   -99.85%  (p=0.000 n=10+9)
ColumnarizeMaterialize-24     356kB ± 0%     200kB ± 0%   -43.94%  (p=0.000 n=10+9)

name                       old allocs/op  new allocs/op  delta
Columnarize-24                19.5k ± 0%      0.0k ± 0%   -99.98%  (p=0.000 n=10+10)
ColumnarizeMaterialize-24     19.5k ± 0%      0.0k ± 0%   -99.77%  (p=0.000 n=10+10)
```
Additionally, the conversion method no longer returns an error that is
encountered during the datum decoding since every caller of the method
is panicking with the error anyway.

Release justification: low-risk improvement.

Release note: None

**colexec: memory accounting at row granularity in columnarizer**

This commit makes it so that we perform the memory accounting at a row
granularity (rather than a batch granularity) in the columnarizer as
well as the values operator. As a result, we will respect memory limits
much better than previously.

The main idea is that previously we were converting from the row-based
format to the columnar one a column at a time, and now we will perform
the conversion one full row at a time. As expected, this leads to
a performance regression, but given the speedup in the previous commit
and the improvement to the stability of Cockroach, it seems worth it.
```
name                       old time/op    new time/op    delta
Columnarize-24                157µs ± 1%     291µs ± 0%  +85.05%  (p=0.000 n=8+10)
ColumnarizeMaterialize-24     514µs ± 0%     648µs ± 0%  +26.09%  (p=0.000 n=9+9)

name                       old speed      new speed      delta
Columnarize-24             1.02GB/s ± 1%  0.55GB/s ± 0%  -45.96%  (p=0.000 n=8+10)
ColumnarizeMaterialize-24   311MB/s ± 0%   247MB/s ± 0%  -20.69%  (p=0.000 n=9+9)

name                       old alloc/op   new alloc/op   delta
Columnarize-24                 229B ± 1%      375B ± 1%  +63.96%  (p=0.000 n=9+10)
ColumnarizeMaterialize-24     200kB ± 0%     200kB ± 0%   +0.05%  (p=0.000 n=10+10)

name                       old allocs/op  new allocs/op  delta
Columnarize-24                 4.00 ± 0%      7.00 ± 0%  +75.00%  (p=0.000 n=10+10)
ColumnarizeMaterialize-24      44.0 ± 0%      47.0 ± 0%   +6.82%  (p=0.000 n=10+10)
```

This commit makes the corresponding change to the method and file names
to indicate that now the conversion is from a single row into all
vectors at once.

Addresses: #76464.

Release justification: bug fix.

Release note: None

----

The performance impact of both commits combined:
```
name                       old time/op    new time/op    delta
Columnarize-24                441µs ± 0%     292µs ± 1%  -33.80%  (p=0.000 n=9+10)
ColumnarizeMaterialize-24     800µs ± 1%     648µs ± 0%  -18.99%  (p=0.000 n=10+10)

name                       old speed      new speed      delta
Columnarize-24              363MB/s ± 0%   548MB/s ± 1%  +51.07%  (p=0.000 n=9+10)
ColumnarizeMaterialize-24   200MB/s ± 1%   247MB/s ± 0%  +23.44%  (p=0.000 n=10+10)

name                       old alloc/op   new alloc/op   delta
Columnarize-24                157kB ± 0%       0kB ± 1%  -99.76%  (p=0.000 n=10+9)
ColumnarizeMaterialize-24     356kB ± 0%     200kB ± 0%  -43.91%  (p=0.000 n=9+9)

name                       old allocs/op  new allocs/op  delta
Columnarize-24                19.5k ± 0%      0.0k ± 0%  -99.96%  (p=0.000 n=10+10)
ColumnarizeMaterialize-24     19.5k ± 0%      0.0k ± 0%  -99.76%  (p=0.000 n=10+9)
```

yuzefovich committed:
colexec: memory accounting at row granularity in columnarizer

This commit makes it so that we perform the memory accounting at a row
granularity (rather than a batch granularity) in the columnarizer as
well as the values operator. As a result, we will respect memory limits
much better than previously.

The main idea is that previously we were converting from the row-based
format to the columnar one a column at a time, and now we will perform
the conversion one full row at a time. As expected, this leads to
a performance regression, but given the speedup in the previous commit
and the improvement to the stability of Cockroach, it seems worth it.
```
name                       old time/op    new time/op    delta
Columnarize-24                157µs ± 1%     291µs ± 0%  +85.05%  (p=0.000 n=8+10)
ColumnarizeMaterialize-24     514µs ± 0%     648µs ± 0%  +26.09%  (p=0.000 n=9+9)

name                       old speed      new speed      delta
Columnarize-24             1.02GB/s ± 1%  0.55GB/s ± 0%  -45.96%  (p=0.000 n=8+10)
ColumnarizeMaterialize-24   311MB/s ± 0%   247MB/s ± 0%  -20.69%  (p=0.000 n=9+9)

name                       old alloc/op   new alloc/op   delta
Columnarize-24                 229B ± 1%      375B ± 1%  +63.96%  (p=0.000 n=9+10)
ColumnarizeMaterialize-24     200kB ± 0%     200kB ± 0%   +0.05%  (p=0.000 n=10+10)

name                       old allocs/op  new allocs/op  delta
Columnarize-24                 4.00 ± 0%      7.00 ± 0%  +75.00%  (p=0.000 n=10+10)
ColumnarizeMaterialize-24      44.0 ± 0%      47.0 ± 0%   +6.82%  (p=0.000 n=10+10)
```

This commit makes the corresponding change to the method and file names
to indicate that now the conversion is from a single row into all
vectors at once.

Release justification: bug fix.

Release note: None

yuzefovich committed:
colexec: remove redundant allocations when converting from rows

This redundant allocation has been present since the introduction of the
vectorized engine (it was introduced in
37957a9b4e6964940ed2dd4d94b4dbfa671a11aa I think by accident - the
impact of some changes in that commit were probably not understood and
we didn't have the benchmarks to catch the regression).
```
Columnarize-24                442µs ± 1%     157µs ± 0%   -64.49%  (p=0.000 n=10+9)
ColumnarizeMaterialize-24     801µs ± 1%     513µs ± 0%   -35.89%  (p=0.000 n=10+10)

name                       old speed      new speed      delta
Columnarize-24              362MB/s ± 1%  1019MB/s ± 0%  +181.61%  (p=0.000 n=10+9)
ColumnarizeMaterialize-24   200MB/s ± 1%   312MB/s ± 0%   +55.97%  (p=0.000 n=10+10)

name                       old alloc/op   new alloc/op   delta
Columnarize-24                157kB ± 0%       0kB ± 1%   -99.85%  (p=0.000 n=10+9)
ColumnarizeMaterialize-24     356kB ± 0%     200kB ± 0%   -43.94%  (p=0.000 n=10+9)

name                       old allocs/op  new allocs/op  delta
Columnarize-24                19.5k ± 0%      0.0k ± 0%   -99.98%  (p=0.000 n=10+10)
ColumnarizeMaterialize-24     19.5k ± 0%      0.0k ± 0%   -99.77%  (p=0.000 n=10+10)
```
Additionally, the conversion method no longer returns an error that is
encountered during the datum decoding since every caller of the method
is panicking with the error anyway.

Release justification: low-risk improvement.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87343)
<!-- Reviewable:end -->

msirek reviewed (approved):
The fix is useful for backporting.  I don't know if you'd like to wait a week or two before merging since one of the commits impacts performance (in case the other commit doesn't offset it in all cases). Or, could merge now and revert if there's an issue in performance tests.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 6 of 6 files at r1, 17 of 17 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87343#-:-NAzLDaU1Fwc0ruoGifu:b-tn4stc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Yeah, I plan to wait a week or so before merging, just to let it bake on master a bit.

yuzefovich committed:
colbuilder: fix invalid sharing of a memory account by the columnarizer

We recently merged a couple of changes to the memory accounting in the
columnarizer which relied on its memory account not being shared with
any other component. This, however, wasn't true since we passed the
"streaming" memory account that many components can use. This commit
fixes that oversight by always creating a separate account.

The previous behavior could result in the columnarizer clearing the
account when another user of that account attempts to shrink it (in
production builds this would only trigger a sentry report with no panic).
For example, this situation could occur when we plan a vectorized
ordered aggregator that uses a default aggregate function and then we
handle the rendering with a wrapped row-by-row processor which requires
a creation of the columnarizer.

Release justification: bug fix.

Release note: None (no release with this bug)

yuzefovich commented:
I included #87471 into this PR. It seems like 22.1.7 should be picking the SHA soon, and I want for this change to bake for some more time, so we'll target 22.1.8 for this patch.
